src/writer.rs

```diff
@@ -169,20 +169,134 @@ impl<W: Write> Writer<W> {
+    /// use std::io::Cursor;
+    ///
+    /// let mut writer = Writer::new(Cursor::new(Vec::new()));
+    ///
+    /// // writes <tag attr1="value1"/>
+    /// writer.create_element("tag")
+    ///     .with_attribute(("attr1", "value1"))  // chain `with_attribute()` calls to add many attributes
+    ///     .write_empty()?;
+    ///
+    /// // writes <tag attr1="value1" attr2="value2">with some text inside</tag>
+    /// writer.create_element("tag")
+    ///     .with_attributes(vec![("attr1", "value1"), ("attr2", "value2")].into_iter())  // or add attributes from an iterator
+    ///     .write_text_content(BytesText::from_plain_str("with some text inside"))?;
+    ///
+    /// // writes <tag><fruit quantity="0">apple</fruit><fruit quantity="1">orange</fruit></tag>
+    /// writer.create_element("tag")
+    ///     .write_nested_content(|writer| {
+    ///         let fruits = ["apple", "orange"];
+    ///         for (quant, item) in fruits.iter().enumerate() {
+    ///             writer
+    ///                 .create_element("fruit")
+    ///                 .with_attribute(("quantity", quant.to_string().as_str()))
+    ///                 .write_text_content(BytesText::from_plain_str(item))?;
+    ///         }
+    ///         Ok(())
+    ///     })?;
+    /// # Ok(())
+    /// # }
     /// ```
-    pub fn write_nested_event<'a, N: AsRef<[u8]>, E: AsRef<Event<'a>>>(
-        &mut self,
-        parent_name: N,
-        event: E,
-    ) -> Result<()> {
-        let name = parent_name.as_ref();
-        self.write_event(Event::Start(BytesStart::borrowed(name, name.len())))?;
-        self.write_event(event)?;
-        self.write_event(Event::End(BytesEnd::borrowed(name)))?;
-        Ok(())
+    #[must_use]
+    pub fn create_element<'a, N>(&'a mut self, name: &'a N) -> ElementWriter<W>
+    where
+        N: 'a + AsRef<[u8]> + ?Sized,
+    {
+        ElementWriter {
+            writer: self,
+            start_tag: BytesStart::borrowed_name(name.as_ref()),
+        }
+    }
+}
+
+pub struct ElementWriter<'a, W: Write> {
+    writer: &'a mut Writer<W>,
+    start_tag: BytesStart<'a>,
+}
+
+impl<'a, W: Write> ElementWriter<'a, W> {
+    /// Adds an attribute to this element.
+    pub fn with_attribute<'b, I>(mut self, attr: I) -> Self
+    where
+        I: Into<Attribute<'b>>,
+    {
+        self.start_tag.push_attribute(attr);
+        self
+    }
+
+    /// Add additional attributes to this element using an iterator.
+    ///
+    /// The yielded items must be convertible to [`Attribute`] using `Into`.
+    ///
+    /// [`Attribute`]: attributes/struct.Attributes.html
+    pub fn with_attributes<'b, I>(mut self, attributes: I) -> Self
+    where
+        I: IntoIterator,
+        I::Item: Into<Attribute<'b>>,
+    {
+        self.start_tag.extend_attributes(attributes);
+        self
+    }
+
+    /// Write some text inside the current element.
+    pub fn write_text_content(self, text: BytesText) -> Result<&'a mut Writer<W>> {
```

Comment:
Does a pair of tags with no text always get represented as [Start] [Text ""] [End]?

Reply:
What do you mean? If possible I think we could *optimize* it and write an empty event but this is really not very important.

Reply:
I just wanted to be sure that what I'm doing is correct - if it was ever necessary to write two separate start and end events with no text event (even an empty one) in the middle, this wouldn't be convenient. But if there needs to always be one, then this is fine.